crates/polkavm-linker/src/program_from_elf.rs

```
@@ -6399,7 +6399,7 @@ fn emit_code(
                             if is_optimized {
                                 unreachable!("internal error: instruction with only constant operands: {op:?}")
                             } else {
-                                I::load_imm(dst, OperationKind::from(kind).apply_const(src1 as i32, src2 as i32) as u32)
+                                I::load_imm(dst, OperationKind::from(kind).apply_const(i64::from(src1), i64::from(src2)) as u32)
```

Do not use truncating casts that silently convert the types. If we have a codegen bug (as in - if we have an immediate argument that needs more than 32-bits) we want the code to panic if it cannot fit, instead of silently generating wrong code.

---

Hm, well, here `src1` and `src2` both are 32bit, so conversion would always happen.

But I think, you meant the intermediate representation should be 64bit...

---

Operands are coming from `RegImm`, which we are not suppose to change for the while we have 32 bit support intact.

---

> Operands are coming from `RegImm`, which we are not suppose to change for the while we have 32 bit support intact.

No, they are not supposed to *ever* change, because even if we go full 64-bit only we won't be able to encode 64-bit immediates for every instruction. (:

Anyway, it doesn't matter where they are coming from. If there is *ever* a value that needs more than 32-bits here then this code is *totally broken*. This might not be the case right now, but this file is already a big ball of complex code and mud. I don't want to have to manually make sure this invariant is true and keep it in mind every time I change anything. So we want what is essentially an assertion that *if* by accident we get a too big of a value this code should panic and let us know that it's broken. That's why I said this should not be an infallible, truncating cast.